integration/kotlinx-coroutines-play-services/src/Tasks.kt

```diff
@@ -91,17 +136,28 @@ public suspend fun <T> Task<T>.await(): T {
         } else {
             throw e
         }
+    } else if (cancellationTokenSource.token.isCancellationRequested) {
```

Comment:
I decided here to check for the cancellation of the `CancellationTokenSource` after checking for if the `Task` was already completed. I could also see the argument for going the other way instead, where we throw `CancellationException` if the `cancellationTokenSource` is cancelled, even if the `Task` is already completed.

Reply:
Doesn't feel like this check should be present at all.

Consider the following scenario:
1) `isComplete` returned false
2) Task failed with an application-specific exception
3) Concurrently the cancellation has been requested.
4) `isCancellationRequested` returned `true

In such scenario, we will throw `CancellationException` instead of application-specific exception and it will effectively became lost

Reply:
Ah, that makes sense for the race

The reason why I was checking for the token being cancelled directly was to guard against a scenario where the `cancellationTokenSource` is cancelled but the `Deferred` will wait forever, if the `cancellationTokenSource` doesn't control the `Task`:

```kotlin
val cancellationTokenSource = CancellationTokenSource()
// Construct a task without the cancellation token source
val taskCompletionSource = TaskCompletionSource<Int>()

val deferred: Deferred<Int> = async(start = CoroutineStart.LAZY) {
    taskCompletionSource.task.await(cancellationTokenSource)
}

assertFalse(deferred.isCompleted)
cancellationTokenSource.cancel()

try {
    deferred.await() // will suspend forever
    fail("deferred.await() should be cancelled")
} catch (e: Exception) {
    assertTrue(e is CancellationException)
}
```

I will remove those checks to avoid the race condition, and explicitly note that the `CancellationTokenSource` should only be used to create the task, and not an external method of cancellation.